OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseOutlineFragment.java

```diff
@@ -387,32 +384,70 @@ private void restore(@Nullable Bundle savedInstanceState) {
     }
 
     private void fetchCourseComponent() {
-        final String courseId = courseData.getCourseId();
-        if (courseComponentId != null && TextUtils.isEmpty(screenName)) {
-            final CourseComponent courseComponent = courseManager.getComponentByIdFromAppLevelCache(courseId, courseComponentId);
-            if (courseComponent != null) {
-                // Course data exist in app session cache
-                loadData(courseComponent);
-                return;
-            }
-        }
-        // Check if course data is available in app session cache
-        final CourseComponent courseComponent = courseManager.getCourseDataFromAppLevelCache(courseId);
-        if (courseComponent != null) {
-            // Course data exist in app session cache
-            loadData(courseComponent);
-            return;
-        }
-        // Check if course data is available in persistable cache
-        loadingIndicator.setVisibility(View.VISIBLE);
+        initCourseObservers();
         // Prepare the loader. Either re-connect with an existing one or start a new one.
         if (environment.getLoginPrefs().isUserLoggedIn()) {
-            getLoaderManager().initLoader(0, null, this);
+            final String courseId = courseData.getCourseId();
+            courseViewModel.getCourseData(courseId, courseComponentId, true, false,
+                    CourseViewModel.CoursesRequestType.CACHE.INSTANCE);
         } else {
             EventBus.getDefault().post(new LogoutEvent());
         }
     }
 
+    private void initCourseObservers() {
+        courseViewModel = new ViewModelProvider(this).get(CourseViewModel.class);
+        courseViewModel.getCourseComponent().observe(getViewLifecycleOwner(), new EventObserver<>(courseComponent -> {
+            if (!isAdded()) {
+                return null;
+            }
+            loadData(validateCourseComponent(courseComponent));
+            FullscreenLoaderDialogFragment fullscreenLoader = FullscreenLoaderDialogFragment
+                    .getRetainedInstance(getChildFragmentManager());
+            if (fullscreenLoader != null && fullscreenLoader.isResumed()) {
+                new SnackbarErrorNotification(listView).showUpgradeSuccessSnackbar(R.string.purchase_success_message);
+                fullscreenLoader.closeLoader();
+            }
+            return null;
+        }));
+
+        courseViewModel.getShowProgress().observe(getViewLifecycleOwner(), showProgress -> {
+            loadingIndicator.setVisibility(showProgress ? View.VISIBLE : View.GONE);
```

Comment:
The loading indicator never stops when we open a non-cached course with no internet.